sync/sync.go

```diff
@@ -0,0 +1,353 @@
+	GetId() uint32
+}
+
+type Layer interface {
+	Index() int
+	Blocks() []Block
+	Hash() string
+}
+
+type BlockValidator interface {
+	ValidateBlock(block Block) bool
+}
+
+type Configuration struct {
+	hdist          uint32 //dist of consensus layers from newst layer
+	syncInterval   time.Duration
+	concurrency    int //number of workers for sync method
+	requestTimeout time.Duration
+}
+
+type Syncer struct {
+	peers     Peers
+	layers    mesh.Mesh
+	sv        BlockValidator //todo should not be here
+	config    Configuration
+	p         *p2p.Protocol
+	SyncLock  uint32
+	startLock uint32
+	forceSync chan bool
+	exit      chan struct{}
+}
+
+func (s *Syncer) ForceSync() {
+	s.forceSync <- true
+}
+
+func (s *Syncer) Close() {
+	close(s.forceSync)
+	close(s.exit)
+}
+
+const (
+	IDLE       uint32          = 0
+	RUNNING    uint32          = 1
+	BLOCK      p2p.MessageType = 1
+	LAYER_HASH p2p.MessageType = 2
+	LAYER_IDS  p2p.MessageType = 3
+
+	protocol = "/sync/fblock/1.0/"
+)
+
+func (s *Syncer) IsSynced() bool {
+	return s.layers.LocalLayerCount() == s.maxSyncLayer()
+}
+
+func (s *Syncer) Stop() {
+	s.exit <- struct{}{}
+}
+
+func (s *Syncer) Start() {
+	if atomic.CompareAndSwapUint32(&s.startLock, 0, 1) {
+		go s.run()
+		return
+	}
+}
+
+func NewSync(peers Peers, layers mesh.Mesh, bv BlockValidator, conf Configuration) *Syncer {
+	s := Syncer{peers, layers, bv, conf, p2p.NewProtocol(peers, protocol, time.Second*5), 0, 0, make(chan bool), make(chan interface{})}
+	s.p.RegisterMsgHandler(LAYER_HASH, s.LayerHashRequestHandler)
+	s.p.RegisterMsgHandler(BLOCK, s.BlockRequestHandler)
+	s.p.RegisterMsgHandler(LAYER_IDS, s.LayerIdsRequestHandler)
+	return &s
+}
+
+//fires a sync every sm.config.syncInterval or on force space from outside
+func (s *Syncer) run() {
+	syncTicker := time.NewTicker(s.config.syncInterval)
+	for {
+		doSync := false
+		select {
+		case <-s.exit:
+			log.Debug("run stoped")
+			return
+		case doSync = <-s.forceSync:
+		case <-syncTicker.C:
+			doSync = true
+		default:
+			doSync = false
+		}
+		if doSync {
+			go func() {
+				if atomic.CompareAndSwapUint32(&s.SyncLock, IDLE, RUNNING) {
+					log.Debug("do sync")
+					s.Synchronise()
+					atomic.StoreUint32(&s.SyncLock, IDLE)
+				}
+			}()
+		}
+	}
+}
+
+func (s *Syncer) maxSyncLayer() uint32 {
+	if uint32(s.layers.LatestKnownLayer()) < s.config.hdist {
+		return 0
+	}
+
+	return s.layers.LatestKnownLayer() - s.config.hdist
+}
+
+func (s *Syncer) Synchronise() {
+	for i := s.layers.LocalLayerCount(); i < s.maxSyncLayer(); {
+		i++
+		blockIds := s.GetLayerBlockIDs(i) //returns a set of all known blocks in the mesh
+		ids := make(chan uint32, len(blockIds))
+		output := make(chan Block)
+
+		//bufferd chan for workers  to read from
+
+		for _, id := range blockIds {
+			ids <- id
+		}
+
+		close(ids)
+
+		// each worker goroutine tries to fetch a block iteratively from each peer
+		for i := 0; i < s.config.concurrency; i++ {
+			go func() {
+				for id := range ids {
```

Comment:
range ids in a multi-reader action, right? meaning that each worker will query all peers for all blocks. Is it what it's suppose to do?

Reply:
Also, 'ids' can contain duplicate blocks. Consider removing them before querying neighbours